source/non_matching/fe_values.cc

```diff
@@ -0,0 +1,523 @@
+        Assert(cell_is_set(), ExcCellNotSet());
+
+        double value = 0;
+        for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+          value += local_dof_values[i] *
+                   element->shape_value_component(i, point, component);
+
+        return value;
+      }
+
+
+
+      template <int dim, class VectorType>
+      Tensor<1, dim>
+      RefSpaceFEFieldFunction<dim, VectorType>::gradient(
+        const Point<dim> & point,
+        const unsigned int component) const
+      {
+        AssertIndexRange(component, this->n_components);
+        Assert(cell_is_set(), ExcCellNotSet());
+
+        Tensor<1, dim> gradient;
+        for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+          gradient += local_dof_values[i] *
+                      element->shape_grad_component(i, point, component);
+
+        return gradient;
+      }
+
+
+
+      template <int dim, class VectorType>
+      SymmetricTensor<2, dim>
+      RefSpaceFEFieldFunction<dim, VectorType>::hessian(
+        const Point<dim> & point,
+        const unsigned int component) const
+      {
+        AssertIndexRange(component, this->n_components);
+        Assert(cell_is_set(), ExcCellNotSet());
+
+        Tensor<2, dim> hessian;
+        for (unsigned int i = 0; i < local_dof_indices.size(); ++i)
+          hessian += local_dof_values[i] *
+                     element->shape_grad_grad_component(i, point, component);
+
+        return symmetrize(hessian);
+      }
+    } // namespace FEValuesImplementation
+  }   // namespace internal
+
+
+
+  template <int dim>
+  template <class VectorType>
+  FEValues<dim>::FEValues(const hp::FECollection<dim> &fe_collection,
+                          const Quadrature<1> &        quadrature,
+                          const RegionUpdateFlags      region_update_flags,
+                          const MeshClassifier<dim> &  mesh_classifier,
+                          const DoFHandler<dim> &      dof_handler,
+                          const VectorType &           level_set,
+                          const AdditionalData &       additional_data)
+    : mapping_collection(&dealii::hp::StaticMappingQ1<dim>::mapping_collection)
+    , fe_collection(&fe_collection)
+    , q_collection_1D(quadrature)
+    , region_update_flags(region_update_flags)
+    , mesh_classifier(&mesh_classifier)
+    , quadrature_generator(q_collection_1D, additional_data)
+    , reference_space_level_set(
+        std::make_unique<internal::FEValuesImplementation::
+                           RefSpaceFEFieldFunction<dim, VectorType>>(
+          dof_handler,
+          level_set))
+  {
+    // Tensor products of each quadrature in q_collection_1D. Used on the
+    // non-intersected cells.
+    hp::QCollection<dim> q_collection;
+    for (unsigned int i = 0; i < q_collection_1D.size(); ++i)
+      q_collection.push_back(Quadrature<dim>(q_collection_1D[i]));
+
+    init(q_collection);
```

Comment:
I think we most commonly spell out the full word for this function name, `initialize`, so I would lean towards that name unless we have reasons to stay with this abbreviation (as opposed to `reinit`).